### ISSUE TITLE
PyOpenSSL X509Store / SSLContext parity in Cryptography

### ISSUE THREAD

**Issue by vEpiphyte** [pyca/cryptography#10393]
Hello!

I've been working on updating some code to utilize cryptography in favor of PyOpenSSL due to the API deprecation in the older project.

The only code that I can not currently remove is related to the use of X509Store and X509StoreContext. That is utilized for doing certificate validation. For example:

```
    def verifyACertificate(self, cert: c_x509.Certificate):
    	'''
    	Check if a Certificate is trusted by a set of CAs and is not revoked.
    	'''
        crls = self.getCaCrls()  # type: List[c_x509.CertificateRevocationList]
        cacerts = self.getCaCerts()  # type: List[c_x509.Certificate]

        store = crypto.X509Store()  # This is pyopenssl...
        [store.add_cert(crypto.X509.from_cryptography(cacert)) for cacert in cacerts]

        if crls:
            # Setting flags is important. Other uses use things such as PARTIAL_CHAIN flags.
            store.set_flags(crypto.X509StoreFlags.CRL_CHECK | crypto.X509StoreFlags.CRL_CHECK_ALL)
            [store.add_crl(crypto.CRL.from_cryptography(crl)) for crl in crls]

        ctx = crypto.X509StoreContext(store, crypto.X509.from_cryptography(cert))
        try:
            ctx.verify_certificate()  # raises X509StoreContextError if unable to verify
        except crypto.X509StoreContextError as e:
            mesg = _unpackContextError(e)  # helper function to unpack the X509StoreContextError, not important here.  
            raise Exception(mesg)
        return cert
```

I believe my use case aligns with https://github.com/pyca/cryptography/issues/10276 ( doing code signing and/or user cert verification ). Current docs for the verification APIS ( https://cryptography.io/en/42.0.2/x509/verification/ ) don't seem to support setting CRLs or flag setting.

Is this type of use case in scope for work in https://github.com/pyca/cryptography/pull/10345 ?

**Comment by alex** [comment]
cc: @woodruffw 

Looks like there's two pieces you need here:

1. Doing verification without checking for a particular SAN, which is what @woodruffw has in his PR
2. Checking CLRs.

Both of these are tracked in https://github.com/pyca/cryptography/issues/10034

Figuring out CLRs is going to require some API design work on our part. It looks like your implementation relies on having pre-fetched the CRLs, and not loading them on-demand.

**Comment by vEpiphyte** [comment]
@alex Correct - I am assuming that CRLs are present ahead of time ( and very much not assuming that it is the job of cryptography to retrieve them ).

If these pieces are being correctly tracked in #10034 we can close this issue out to avoid duplication.

**Comment by alex** [comment]
https://github.com/pyca/cryptography/pull/10345 adds verification without a subject.

CRL is the remaining piece here. We still need to figure out what we want to do in terms of API design there.

**Comment by alex** [comment]
@reaperhulk and I discussed some and here's what we came up for in terms of an API:

- We add a new `RevocationChecker` ABC, which has an `is_revoked()` method
- We add a new `revocation_checker` API on `PolicyBuilder` to provide a `RevocationChecker` for use in that verification
- The core verification engine calls `is_revoked` in the right places
- We introduce a `CRLRevocationChecker` that takes a list of `CertificateRevocationList` objects and implements the `RevocationChecker` interface using them

Then in terms of the `CRLRevocationChecker`:

- It needs to verify the signature on the CRL with the issuer's public key
- TODO: decide whether to fail open or closed if a CRL's `next_update` is in the past
- It needs to check the CRL's extensions and make sure there are not any critical extensions that we don't check.
- Then it needs to check if the certificate is found in the CRL
- And then it needs to make sure there are no unknown critical extensions in the CRL entry

And for all of this we need to write a bunch of new x509-limbo tests :-)


For folks who were interested in this feature, does this sound like a design that works for your use case?

@woodruffw do you have an opinion on adding CRL tests to x509-limbo?

**Comment by vEpiphyte** [comment]
That design sounds solid from a user perspective. With the schism between pyopenssl removing the CRL constructs, and older versions having a `cryptography` version constraint which prevents installing `v44.0.1`, this is an important item to nail down IMO.

**Comment by woodruffw** [comment]
Neat -- I realized I forgot to respond before, but I'm a big +1 on adding CRL tests to x509-limbo! They'd definitely make sense under a new `crl::` namespace, similar to what we do for `rfc5280::` and `webpki::`.

**Comment by nbastin** [comment]
I was doing some related work and stumbled across #12780, which is closed so I thought I'd at least add a comment here.  Duplicate certificate serial numbers in CRLs are often legitimate, esp. for issuers that have occasion to use the `toBeRevoked` extension, but also for issuers who have dynamically changed their partitioning scheme at some point or whose deltas are issued by iCRL points that may also have shifted.  Also of course CRL issuance authorities can be authoritative for multiple CAs and are not required to partition their CRLs by CA, so (as stated in the X.509 spec itself) serial number alone does not uniquely identify a certificate.

**Comment by alex** [comment]
On further review I can't actually find the quoted language in RFC 5280 (might be PEBKAC though!). To the extent RFC5280 prohibits them, it's hard to say that they're legitimate! But if they're not actually prohibited by RFC 5280, we should respect them of course.

**Comment by woodruffw** [comment]
I also tried to find matching standards language for this, and IMO RFC 5280 is unfortunately pretty ambiguous here:

1. Serial numbers are expected to be unique per-issuer ([SS 4.1.2.2](https://datatracker.ietf.org/doc/html/rfc5280#section-4.1.2.2))
2. A "full and complete" CRL lists all unexpired certs issued by _a_ CA that have been revoked for any reason ([SS 5](https://datatracker.ietf.org/doc/html/rfc5280#section-5))

From these it follows that a CRL's revoked serials should be unique, since the CRL itself provides a single issuer and SNs are unique per-issuer.

However, the RFC then introduces ambiguity by mentioning "indirect CRLs," for which the issue relationship would be one-many (one CRL issuer, many certificate issuers).

TL;DR: It's unclear from the standards how uniqueness and indirect CRLs interact -- the standard more or less asserts that uniqueness is required, and then defines a PKI scenario where uniqueness can't be meaningfully established. I think in practice most X.509 path validations address this by not supporting indirect CRLs at all (ref. [this](https://security.stackexchange.com/questions/242185/deployment-tips-to-support-indirect-crls)).

**Comment by nbastin** [comment]
CRLs have a single issuer, but that issuer is _not the certificate issuer_.  The `CertificateListContent.issuer` is the issuer of the CRL, _not the certificates in the CRL_.  In the case of indirect CRL authorities, the issuer in the CLC is the iCRL maintainer, and each certificate has a `certificateIssuer` CRL entry extension which indicates who issued the actual certificate in question.

Edit: `certificateIssuer` actually shows up in a bunch of places in the tree of the CRL in apparently an attempt to avoid a lot of repeated information, so the common place for the issuer to be placed in an iCRL is probably in `revokedGroup` (which has an optional `certificateIssuer` component) (or in `ToBeRevokedGroup` for that type).

**Comment by woodruffw** [comment]
> CRLs have a single issuer, but that issuer is not the certificate issuer

Right, I'm not saying it is -- I'm saying that RFC 5280 says that a CRL has _a_ issuer and lists all certs for _a_ CA, so the uniqueness property is present regardless of whether the issuer itself varies. The only way the uniqueness property can break is if there's more than one CA, which is where I think the RFC is ambiguous (since it talks about CRLs as if they have both one relevant CA _and_ mentions iCRLs).

> In the case of indirect CRL authorities, the issuer in the CLC is the iCRL maintainer, and each certificate has a certificateIssuer CRL entry extension which indicates who issued the actual certificate in question.

Okay -- so, if I understand this correctly, the uniqueness relationship would then be keyed on each CRL entry's `certificateIssuer`? So there's still a uniqueness property with SNs per RFC 5280, it's just not as simple as "all listed CRL entries are flatly unique."

**Comment by nbastin** [comment]
Yeah I won't keep editing my previous comment (ignore the part about the extension - it's actually an optional component inside a bunch of other CRL pieces, where each instance scopes anything following it until another one is found), but each certificate is _somewhere_ scoped to a `certificateIssuer` (falling back to the CRL issuer if there's no other option, which of course is the case for a CA that issues its own CRLs).

Although, the extension _does_ exist, so would probably have to also be handled.

**Comment by nbastin** [comment]
The relevant paragraphs from 7.12.2 in 9594-8:2020 probably make more sense than the hash I've made of it in my comments (sorry!):

```
Since the iCRL is authoritative for the CA(s) other that the CRL issuer, serial number alone in the CRL entry does not
uniquely identify a public-key certificate that has been revoked. You also need to identify the CA that issued the public-
key certificate placed on the iCRL. This is achieved by adding the certificateIssuer CRL entry extension (see clause
9.6.2.3). This extension shall always be flagged as critical to ensure that the relying parties process it and associate the
CRL entry with the appropriate CA.

If each entry on a CRL contained the certificateIssuer extension (which is a directory distinguished name), it would
make the CRL size large. Thus, in order to reduce the CRL size, the iCRL issuing authority should sort the CRL entries
by issuing CAs. Using this approach, only the first public-key certificate appearing on the CRL for a given CA needs to
contain the certificateIssuer extension. All subsequent entries are assumed to be for the same public-key certificate
issuing CA until another certificateIssuer CRL entry extension is encountered. To further reduce the size of the
CRL, if the iCRL issuing authority is a CA, it should contain its revoked public-key certificate first, obviating the need
for certificateIssuer extension for any of its certificates.```

**Comment by nbastin** [comment]
Also just to go back to the uniqueness point - you could still potentially find the same cert named in a CRL more than once if it shows up in `ToBeRevoked` and then you update with a delta CRL after the `revocationTime` specified.  The delta (properly formed) will have the cert show up once with reason `removeFromCRL` (for the `toBeRevoked` entry) and then again as a real CRL entry (if it was actually revoked after the time expired).

I suspect this is also possible for `certificateHold` reasons, although we don't use those so I'm not familiar with the intricacies.
